runtime/compiler/z/codegen/J9S390CHelperLinkage.cpp

```diff
@@ -364,6 +373,26 @@ TR::Register * TR::S390CHelperLinkage::buildDirectDispatch(TR::Node * callNode,
       }
    TR::SymbolReference * callSymRef = callNode->getSymbolReference();
    void * destAddr = callNode->getSymbolReference()->getSymbol()->castToMethodSymbol()->getMethodAddress();
+
+#ifdef J9ZOS390
+   if (environment != 0)
+      {
+      // XPLINK GPR5 is an environment register.
+      // Currently this is only used by IFA switching call
+      if (TR::Compiler->target.is32Bit())
+         {
+         cursor = generateRILInstruction(cg(), TR::InstOpCode::LLILF, callNode,
+                                         javaStackPointerRegister, static_cast<int32_t>(environment));
+         }
+      else
+         {
+         int32_t envHi = static_cast<int32_t>((environment >> 32) & 0x00000000FFFFFFFFull);
+         int32_t envLo = static_cast<int32_t>(environment & 0x00000000FFFFFFFFull);
+         cursor = generateRILInstruction(cg(), TR::InstOpCode::IIHF, callNode, javaStackPointerRegister, envHi);
+         cursor = generateRILInstruction(cg(), TR::InstOpCode::IILF, callNode, javaStackPointerRegister, envLo);
+         }
+      }
```

Comment:
This whole sequence should simply use `genLoadAddressConstant` and pass the `environment` constant through. The function will figure out the best sequence to generate.

Reply:
@NigelYiboYu Loading environment register is necessary for calling a C function from JIT code. I omitted that part from Linkage as Assembly listing for `fast_jitInstanceOf` was using it as volatile and we overwrite the register so anyway loaded value was not used. I checked other dual mode helper, Some of them returns the function pointer of the slow path helper function which belongs to same library and in that scenario environment register is needed to get the function pointer which is currently taken care by assembly glue in `znathelp.m4`.  That is the only use I could find currently. \
Correct change here would be respecting XPLINK linkage and also pass environment register for instanceOf as well. Means before calling fastPath helper, environment register should be loaded.